sklearn/cluster/_hdbscan/_linkage.pyx

```diff
@@ -110,23 +109,23 @@ cpdef cnp.ndarray[MST_edge_t, ndim=1, mode='c'] mst_from_data_matrix(
     """
 
     cdef:
-        cnp.int8_t[::1] in_tree
-        cnp.float64_t[::1] min_reachability
-        cnp.int64_t[::1] current_sources
+        uint8_t[::1] in_tree
+        float64_t[::1] min_reachability
+        int64_t[::1] current_sources
         cnp.ndarray[MST_edge_t, ndim=1, mode='c'] mst
 
-        cnp.int64_t current_node, source_node, right_node, left_node, new_node, next_node_source
-        cnp.int64_t i, j, n_samples, num_features
+        int64_t current_node, source_node, new_node, next_node_source
+        int64_t i, j, n_samples, num_features
 
-        cnp.float64_t current_node_core_dist, new_reachability, mutual_reachability_distance
-        cnp.float64_t next_node_min_reach, pair_distance, next_node_core_dist
+        float64_t current_node_core_dist, new_reachability, mutual_reachability_distance
+        float64_t next_node_min_reach, pair_distance, next_node_core_dist
 
-    n_samples = raw_data.shape[0]
-    num_features = raw_data.shape[1]
+    n_samples = len(raw_data)
```

Comment:
I think `.shape[0]` should still work. What is the reason for using `len` here?

Reply:
I ran into the following error, which I had (potentially incorrectly) attributed to the new numpy API:
```python-traceback

sklearn/cluster/_hdbscan/_tree.c: In function '__pyx_f_7sklearn_7cluster_8_hdbscan_5_tree_max_lambdas':
sklearn/cluster/_hdbscan/_tree.c:6453:39: error: 'PyArrayObject' {aka 'struct tagPyArrayObject'} has no member named 'dimensions'
 6453 |   __pyx_t_11 = (__pyx_v_condensed_tree->dimensions[0]);
      |                                       ^~
```

Reply:
For memoryviews, you can still use `.shape`. In this case, `raw_data` is a memoryview so it should still work.

For `cnp.ndarray`, you will have to avoid using `.shape`. I usually recommend going with `PyArray_SHAPE`.

Reply:
Fixed